### PR TITLE
Dependencies

### PR DESCRIPTION
**Pull request by tlevine**
Install development dependencies with "pip install .[dev]".

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -67,4 +67,14 @@
...
                                         'data/*.npy.z']},
           packages=['joblib', 'joblib.test', 'joblib.test.data'],
+          extras_require={
+              'dev': [
+                  'nose>=1.3.7', 'Sphinx>=1.4.8', 'numpydoc>=0.0.6',
+                  'wheel', '.[optional]',
+              ],
+              'optional': [
+                  'numpy>=1.6.1.',
+                  'matplotlib>=1.5.3',
```
Comment (GaelVaroquaux) [inline setup.py]: Why do we need matplotlib?

**Comment by tlevine** [comment]
I don't know why we need matplotlib; the tests gave me errors until I
added it. Maybe that was for the doc directory though.



A note on systax

If you have already downloaded the source code, you can do this.

    pip install '/path/to/source/[optional]'

For example, you can do this.

    pip install '.[optional]'

If you want to install from PyPI, you should do this.

    pip install 'joblib[optional]'

Gael Varoquaux writes:

**Comment by GaelVaroquaux** [comment]
> I don't know why we need matplotlib; the tests gave me errors until I added it.

What errors? Would you mind pasting them here?

**Comment by lesteve** [comment]
I think this is due to the numpydoc sphinx extension which has the matplotlib plot directive.

I would say we should just tell to run the tests via `nosetests joblib` and this problem disappear.

I am not a big fan of adding the extras_require I have to admit, I am not sure they bring any obvious value.

**Comment by tlevine** [comment]
If the documentation does in fact depend on matplotlib matplotlib,
I think we should document that one must install both Sphinx and
matplotlib in order to build the documentation.

But I believe that the documentation does not in fact use the matplotlib
plot directive, as I got similar results when I built the documentation
with matplotlib installed versus without matplotlib installed; I think
we should remove the matplotlib Sphinx extension instead.

I otherwise agree with Loïc's comment.

Loïc Estève writes:

**Comment by GaelVaroquaux** [comment]
> But I believe that the documentation does not in fact use the matplotlib
> plot directive, as I got similar results when I built the documentation
> with matplotlib installed versus without matplotlib installed; I think
> we should remove the matplotlib Sphinx extension instead.

+1 on removing it.

> I otherwise agree with Loïc's comment.

Hum, just to be precise, which one?

**Comment by tlevine** [comment]
This one

> I think this is due to the numpydoc sphinx extension which has the matplotlib
> plot directive.
> 
> I would say we should just tell to run the tests via nosetests joblib and this
> problem disappear.
> 
> I am not a big fan of adding the extras_require I have to admit, I am not sure
> they bring any obvious value.
 
Gael Varoquaux writes:

**Comment by GaelVaroquaux** [comment]
> This one

> > I am not a big fan of adding the extras_require I have to admit, I am not sure
> > they bring any obvious value.

So we are all on the same page: it's pretty much my feeling too.

**Comment by lesteve** [comment]
I edited this one to leave only the nosetests -> nosetests joblib, merging thanks @tlevine.

**Comment by lesteve** [comment]
Rather than tweaking numpydoc it would be a good idea to use napoleon which can parse numpy-style docstring and is included in recent releases of sphinx.

Having said that I am not sure why numpydoc has a plot directive ... which on top of that is only compatible with Python 2.

**Pull request merged**